x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -36,11 +38,19 @@ public class IndicesAccessControl {
     public static final IndicesAccessControl DENIED = new IndicesAccessControl(false, Collections.emptyMap());
 
     private final boolean granted;
-    private final Map<String, IndexAccessControl> indexPermissions;
+    private final Supplier<Map<String, IndexAccessControl>> indexPermissions;
 
     public IndicesAccessControl(boolean granted, Map<String, IndexAccessControl> indexPermissions) {
+        this(granted, () -> Objects.requireNonNull(indexPermissions));
+    }
+
+    public IndicesAccessControl(boolean granted, Supplier<Map<String, IndexAccessControl>> indexPermissionsSupplier) {
         this.granted = granted;
-        this.indexPermissions = Objects.requireNonNull(indexPermissions);
+        this.indexPermissions = Objects.requireNonNull(indexPermissionsSupplier);
```

Comment:
To me it looks clearer if we wrap the Supplier parameter in the `CachedSupplier` type at this stage, rather than delegate this responsability to the caller.